### ISSUE TITLE
LazyTensor cuda fallback

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#62616]
## 🚀 Feature

We currently only have `xla` device to access LazyTensor. Since we are extending LazyTensor support to GPUs, it's not necessary to fallback on CPU when we run into LTC-incompatible operations while we can continue the execution on eager GPU  instead of copying everything back to CPU.

We can start marking `cuda_lazy` and `cpu_lazy` to differentiate tensors' residing device.

**Comment by wconstab** [comment]
We intend the 'LTC_TS_CUDA' env to also cause eager fallback to happen as you described, to cuda not cpu.  There may be a bug.  We'll try to repro and fix, but @jjsjann123 if you can attach a small repro that would help us confirm.  Thanks!

**Comment by jjsjann123** [comment]
I switched to our staging branch and looks like the fallback is going through CUDA for aten operations! awesome!

I do run into issues with custom kernels though.

I have a custom CUDA kernel that assumes inputs on CUDA device. But I'm getting an error
`RuntimeError: Cannot access data pointer of Tensor that doesn't have storage`

Here's a simplified repro:
https://github.com/jjsjann123/ltc_examples/tree/main/custom_kernel_wrong_device

**Comment by desertfire** [comment]
> I switched to our staging branch and looks like the fallback is going through CUDA for aten operations! awesome!
> 
> I do run into issues with custom kernels though.
> 
> I have a custom CUDA kernel that assumes inputs on CUDA device. But I'm getting an error
> `RuntimeError: Cannot access data pointer of Tensor that doesn't have storage`
> 
> Here's a simplified repro:
> https://github.com/jjsjann123/ltc_examples/tree/main/custom_kernel_wrong_device

My understanding is C++ extension does not go through the dispatcher, so it won't trigger LTC and its fallback logic. One solution is to explicitly call input.cuda() in your kernel as you did in your #if branch. @asuhan, any comment?

**Comment by wconstab** [comment]
>C++ extension does not go through the dispatcher
Yes.

> One solution is to explicitly call input.cuda()
I think this should work.

The only other solution I can think of is to implement one of the at::Tensor methods such as data() on LTC, and make it implicitly trigger a sync and copy.  I'm not sure what the considerations might be here.  Was avoiding implementing `data()` a design choice, to make the system more predictable or avoid some edge cases, or is it just something that wasn't yet necessary and maybe harmless to add?

**Comment by asuhan** [comment]
Yes, leaving out `data()` (or `storage()`) is a design choice. Having a graph compiler means that only the inputs and the outputs have to contain data and none of temporaries have to be materialized, should the compiler decide so. Leaving that decision up to the compiler means that looking at the underlying data is meaningless - there could be none. Think of what happens for element-wise fusion of `a + b + c`, there is no storage for `a + b`.

@desertfire I think that would work, yes. I'm thinking we could also add `maybe_data()` / `maybe_storage()` methods to allow zero-copy interop if the lazy tensors already wrap an immediate CUDA tensors. They'd just delegate to `data()` and `storage()` on the wrapped tensor in such cases and return null otherwise.

**Comment by desertfire** [comment]
> @desertfire I think that would work, yes. I'm thinking we could also add `maybe_data()` / `maybe_storage()` methods to allow zero-copy interop if the lazy tensors already wrap an immediate CUDA tensors. They'd just delegate to `data()` and `storage()` on the wrapped tensor in such cases and return null otherwise.

Does that mean if the input to a custom kernel is an expression like, `a+b`, it will still fail?

In the spirit of seamlessly compatible with eager, my suggestion to manually add .cuda() is probably a bad idea.

**Comment by asuhan** [comment]
@desertfire Yes, it should, to avoid surprising performance cliffs in the API. The caller can always decide to materialize `a + b` first, so it's strictly more flexible to have `mabye_` versions then bundle them with materialization in a single API.

Regarding `.cuda()`, yeah, I agreed with it in spirit. You could use `to(device='...')` with CUDA / CPU depending on whether LTC_TS_CUDA is set.

**Comment by desertfire** [comment]
@jjsjann123 , does it make sense if we provide a Python API to wrap around custom kernel invocation? Under the hood, the wrapper would call .to for all the tensor inputs and then call the actual custom kernel.

**Comment by jjsjann123** [comment]
> @jjsjann123 , does it make sense if we provide a Python API to wrap around custom kernel invocation? Under the hood, the wrapper would call .to for all the tensor inputs and then call the actual custom kernel.

So this is transparent to user and there would be no code change for them to switch their models to lazy tensor?

**Comment by desertfire** [comment]
> So this is transparent to user and there would be no code change for them to switch their models to lazy tensor?

In this case, user needs to change their model code. The maybe_data/maybe_storage solution that Alex suggested would not work for all the cases. I don't know if there is a clean solution here that automatically does everything under the hood.

**Comment by jjsjann123** [comment]
So the change is only to decorate custom kernels so we can have the `to` method automatically wrapped on inputs/outputs of custom kernel?
This sounds reasonable to me for the short term. This is not a show blocker that we have to solve it properly right now. We can just use the wrapper as WAR until this becomes a real problem later.

**Comment by wconstab** [comment]
It seems to me that a 'custom extension' is too flexible, that it can never be compatible with a 'general compiler backend', because the compiler can never reason about what might happen inside it.

For the short term I think it's ok to update the custom extension to call .to on the appropriate device.  But the bottom line is, any user code that uses a custom extension is going to be less general than user code that uses the pytorch API.

Longer term, can we do anything better?  Maybe if we change the custom_extension api to make it aware of tensor inputs/outputs, and make it look more like a custom operator.

**Comment by jjsjann123** [comment]
cc'ing @kevinstephano
